src/freertos_drivers/esp_idf/EspIdfWiFi.cxx

```diff
@@ -0,0 +1,1355 @@
+        if ((*it).service_ != service)
+        {
+            // Not a match.
+            continue;
+        }
+
+        // Found a potential match. Check for an address list
+        looking = true;
+        size_t addr_cnt = (*it).addr_.size();
+        if (addr_cnt == 0)
+        {
+            // No resolved addresses to report.
+            continue;
+        }
+
+        constexpr size_t total_size =
+            sizeof(struct addrinfo) + sizeof(struct sockaddr_storage);
+        static_assert(
+            total_size <= NETDB_ELEM_SIZE, "total_size > NETDB_ELEM_SIZE");
+        struct addrinfo *current = *addr;
+        struct addrinfo *last = nullptr;
+
+        for (auto lit = (*it).addr_.begin(); lit != (*it).addr_.end(); ++lit)
+        {
+// The caller is expected to free the struct addrinfo using the method
+// freeaddrinfo. In the lwIP implementation, they use a special buffer pool
+// to free the struct addrinfo to. Therefore, we must also allocate from the
+// same pool.
+#if !defined(LWIP_HDR_MEMP_H)
+#error "lwIP memory pool implementation required and not found."
+#endif
+            current = (struct addrinfo*)memp_malloc(MEMP_NETDB);
+            if (!current)
+            {
+                // Out of memory in the pool
+                break;
+            }
+            memset(current, 0, total_size);
+            if (last)
+            {
+                // Link the last item to us.
+                last->ai_next = current;
+            }
+            else
+            {
+                // This is the first item.
+                *addr = current;
+            }
+            MDNSCacheAddr *ca = &(*lit);
+#if defined(ESP_IDF_WIFI_IPV6)
+            size_t addr_len = AF_INET == ca->family_ ?
+                sizeof(struct sockaddr_in) : sizeof(struct sockaddr_in6);
+            HASSERT(ca->family_ == AF_INET || ca->family_ == AF_INET6);
+#else
+            size_t addr_len = sizeof(struct sockaddr_in);
+            HASSERT(ca->family_ == AF_INET);
+#endif
+            current->ai_family = ca->family_;
+            current->ai_addrlen = sizeof(struct sockaddr_storage);
+            current->ai_addr = (struct sockaddr*)(current + 1);
+            memcpy(current->ai_addr, &ca->addr_, addr_len);
+            last = current;
+            LOG(INFO, "wifi: mdns_lookup() address: %s, port: %u",
+                ipv4_to_string(ntohl(ca->addrIn_.sin_addr.s_addr)).c_str(),
+                ntohs(ca->addrIn_.sin_port));
+        }
+        if (last)
+        {
+            last->ai_next = nullptr;
+        }
+        return *addr ? 0 : EAI_MEMORY;
+    }
+    if (!looking)
+    {
+        // Not currently looking for the service. Start looking.
+        mdns_scan(service);
+    }
+    return EAI_AGAIN;
+}
+
+//
+// EspIdfWiFi::mdns_scan()
+//
+void EspIdfWiFi::mdns_scan(const char *service)
+{
+    bool match = false;
+    OSMutexLock locker(this);
+    for (auto it = mdnsClientCache_.begin(); it != mdnsClientCache_.end(); ++it)
```

Comment:
we should consider STL `find_if` usage here which can return a pointer to the entry if found.